packages/service-clients/odsp-client/src/odspClient.ts

```diff
@@ -89,34 +84,102 @@ function wrapConfigProvider(baseConfigProvider?: IConfigProviderBase): IConfigPr
 	return wrapConfigProviderWithDefaults(baseConfigProvider, odspClientFeatureGates);
 }
 
+/**
+ * Creates OdspClient
+ * @param properties - properties
+ * @returns OdspClient
+ */
+function createOdspClientCore(
+	driverFactory: IDocumentServiceFactory,
+	urlResolver: OdspDriverUrlResolver,
+	connectionConfig: OdspSiteLocation,
+	logger?: ITelemetryBaseLogger,
+	configProvider?: IConfigProviderBase,
+): IOdspClient {
+	return new OdspClient(driverFactory, urlResolver, connectionConfig, logger, configProvider);
+}
+
+/**
+ * Creates OdspClient
+ * @param properties - properties
+ * @returns OdspClient
+ * @alpha
+ */
+export function createOdspClient(properties: OdspClientProps): IOdspClient {
+	return createOdspClientCore(
+		new OdspDocumentServiceFactory(
+			async (options) => getStorageToken(options, properties.connection.tokenProvider),
+			async (options) => getWebsocketToken(options, properties.connection.tokenProvider),
+			properties.persistedCache,
+			properties.hostPolicy,
+		),
+		new OdspDriverUrlResolver(),
+		properties.connection,
+		properties.logger,
+		properties.configProvider,
+	);
+}
+
+/**
+ * Fluid Container type
+ * @alpha
+ */
+export type IOdspFluidContainer<TContainerSchema extends ContainerSchema = ContainerSchema> =
```

Comment:
Adding service-specific IFluidContainer types goes against the whole purpose of a unified IFluidContainer.  I don't really understand your goal - maybe this would be better to start with a conversation.

Reply:
Take a look at IFluidContainer.attach() signature. It's
```ts
attach(props?: unknown): Promise<string>;
```
But OdspClient returns a container that has this signature, and that is not reflected in the signature:

```ts
const attach = async (odspProps?: ContainerAttachProps<OdspContainerAttachProps>): Promise<string> => {
```

This is what is problematic in today's code, even before any changes are made.
But we also need to support more scenarios (see PR description - it describes goals), more custom attach & get flows.

Doing nothing (or worse - keep building on this pattern of a method that has no type defined) seems wrong to me - how customers would know what is supported, and how do they make sure they are catching errors at compile time, not at runtime?

That all said, it's an option. There are other options, like
1. Add IFluidContainerCore (no attach method on it), and service specific containers deriving from IFluidContainerCore. 2. Remove attach() from IFluidContainer and make createContainer() return attach callback in addition to everything it returns today (I think it's worse option).
3. Maybe the outcome - 1st party scenarios could not be accomplished through OdspClient and it's only for very limited subset of 3rd party scenarios.
   - I think we would still need to solve today's attach() signature issue, but it becomes its own independent problem that can be addressed later.

Reply:
Let's chat, I'm not understanding what you're trying to say.

Reply:
Sure, but it's as simple as this: this PR should fail compilation, but it will not:

https://github.com/microsoft/FluidFramework/pull/22144